File: frontend/src/types/checkin.ts
import type { GoalStatus } from "@/types/manager";

export type Quarter = "Q1" | "Q2" | "Q3" | "Q4";
export type ProgressStatus = "not_started" | "on_track" | "completed";

export interface CheckinEmployee {
  id: number;
  email: string;
  full_name: string | null;
}

export interface CheckinGoal {
  id: number;
  employee_id: number;
  title: string;
  description: string | null;
  target: string | null;
  uom: string | null;
  weightage: number;
  status: GoalStatus;
  is_locked: boolean;
  manager_comment: string | null;
  submitted_at: string | null;
  reviewed_at: string | null;
  reviewed_by: number | null;
  approved_at: string | null;
  created_at: string;
  employee: CheckinEmployee | null;
}

export interface ManagerCheckin {
  id: number;
  achievement_id: number;
  manager_comment: string;
  reviewed_by: number;
  reviewed_at: string;
}

export interface AchievementUpdate {
  id: number;
  goal_id: number;
  quarter: Quarter;
  planned_value: string | null;
  actual_value: string | null;
  progress_score: number;
  status: ProgressStatus;
  employee_comment: string | null;
  created_at: string;
  updated_at: string;
  goal: CheckinGoal | null;
  manager_checkins: ManagerCheckin[];
}

export interface GoalListResponse {
  goals: CheckinGoal[];
}

export interface AchievementListResponse {
  updates: AchievementUpdate[];
}

export interface AchievementCreatePayload {
  goal_id: number;
  quarter: Quarter;
  actual_value?: string | null;
  status: ProgressStatus;
  employee_comment?: string | null;
}

export interface AchievementUpdatePayload {
  actual_value?: string | null;
  status: ProgressStatus;
  employee_comment?: string | null;
}

export interface ManagerCheckinPayload {
  manager_comment: string;
}

File: backend/app/schemas/checkin.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.goal import GoalRead

Quarter = Literal["Q1", "Q2", "Q3", "Q4"]
ProgressStatus = Literal["not_started", "on_track", "completed"]


class AchievementCreate(BaseModel):
    goal_id: int
    quarter: Quarter
    actual_value: str | None = Field(default=None, max_length=255)
    status: ProgressStatus = "not_started"
    employee_comment: str | None = Field(default=None, max_length=2000)


class AchievementUpdateRequest(BaseModel):
    actual_value: str | None = Field(default=None, max_length=255)
    status: ProgressStatus
    employee_comment: str | None = Field(default=None, max_length=2000)


class ManagerCheckinCreate(BaseModel):
    manager_comment: str = Field(min_length=1, max_length=2000)


class ManagerCheckinRead(BaseModel):
    id: int
    achievement_id: int
    manager_comment: str
    reviewed_by: int
    reviewed_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AchievementRead(BaseModel):
    id: int
    goal_id: int
    quarter: Quarter
    planned_value: str | None
    actual_value: str | None
    progress_score: float
    status: ProgressStatus
    employee_comment: str | None
    created_at: datetime
    updated_at: datetime
    goal: GoalRead | None = None
    manager_checkins: list[ManagerCheckinRead] = []

    model_config = ConfigDict(from_attributes=True)


class AchievementListResponse(BaseModel):
    updates: list[AchievementRead]


class TeamProgressResponse(BaseModel):
    updates: list[AchievementRead]
